### ISSUE TITLE
Make use of io.Reader 

### ISSUE THREAD

**Issue by Redundancy** [snorristurluson/exsim-bot#2]
https://github.com/snorristurluson/exsim-bot/blob/cecc9a08ed60173c4b6f85fbd5ce4e0a07a77ada/simplebot.go#L53

The connection implements io.Reader, so you can make use of that by relying on the many Go functions that can use it. bufio.NewScanner can give you a scanner that will read and return a full line or json.Decoder will read a stream of json objects (https://golang.org/pkg/encoding/json/#Decoder)

**Comment by snorristurluson** [comment]
Yeah, thanks, will look into that. I was just running into an issue with the json payload getting too large for the read buffer, was just about to look for better ways around that.

**Comment by Redundancy** [comment]
If you check out the example from the json.Decoder docs, it should handle a stream of individual json messages.

In terms of extending that to more than one message type, I recommend having a base message that has a message type field in it, and a payload field. You can use json.RawMessage to defer attempting to fill a struct until you've read the type field. https://github.com/mitchellh/mapstructure can also be useful.
